packages/next-app-router/src/index.ts

```diff
@@ -0,0 +1,75 @@
+'use client';
+import { createUseFunnel } from '@use-funnel/core';
+import { useSearchParams } from 'next/navigation';
+import { useLayoutEffect, useMemo, useState } from 'react';
+
+export const useFunnel = createUseFunnel(({ id, initialState }) => {
+  const searchParams = useSearchParams();
+  const [state, setState] = useState<Record<string, any>>({});
+  useLayoutEffect(() => {
+    if (typeof window !== 'undefined') {
+      setState(window.history.state);
+    }
+
+    function handlePopState(event: PopStateEvent) {
+      setState(event.state);
+    }
+    window.addEventListener('popstate', handlePopState);
+    return () => {
+      window.removeEventListener('popstate', handlePopState);
+    };
+  }, []);
+
+  const currentStep = searchParams.get(`${id}.step`);
+  const currentContext = state?.[`${id}.context`];
+
+  const currentState = useMemo(() => {
+    return currentStep != null && currentContext != null
+      ? ({
+          step: currentStep,
+          context: currentContext,
+        } as typeof initialState)
+      : initialState;
+  }, [currentStep, currentContext, initialState]);
+
+  const history: (typeof initialState)[] = useMemo(
+    () => state?.[`${id}.histories`] ?? [currentState],
+    [state, currentState],
+  );
+
+  const currentIndex = history.length - 1;
+  return useMemo(
+    () => ({
+      history,
+      currentIndex,
+      currentState,
+      push(newState) {
+        const newSearchParams = new URLSearchParams(searchParams);
+        const newHistoryState = {
+          ...state,
+          [`${id}.context`]: newState.context,
+          [`${id}.histories`]: [...(history ?? []), newState],
+        };
+
+        newSearchParams.set(`${id}.step`, newState.step);
+        window.history.pushState(newHistoryState, '', `?${newSearchParams.toString()}`);
+        setState(newHistoryState);
+      },
+      replace(newState) {
+        const newSearchParams = new URLSearchParams(searchParams);
+        newSearchParams.set(`${id}.step`, newState.step);
+        const newHistoryState = {
+          ...state,
+          [`${id}.context`]: newState.context,
+          [`${id}.histories`]: [...(history ?? []), newState],
+        };
+        window.history.replaceState(newHistoryState, '', `?${newSearchParams.toString()}`);
+        setState(newHistoryState);
+      },
+      go(index) {
+        window.history.go(index);
+      },
+    }),
+    [history, currentIndex, currentState, searchParams, id, state],
```

Comment:
```suggestion
    [history, currentIndex, currentState, searchParams, id],
```

I think history.state is not to need `createUseFunnel` result. It seems that `currentIndex`, `currentState`, `history` is enough to construct the result.